File: day6/kotlin/src/Day6.kt
import java.io.File

object Day6 {

    @JvmStatic
    fun main(args: Array<String>) {
        val file = File(javaClass.getResource("day6.txt").toURI())
        val groups = file.readGroups()

        // part1
        val sum = groups.sumBy { it.countAnyContains() }
        println(sum)

        // part2
        val sum2 = groups.sumBy { it.countEachContains() }
        println(sum2)
    }

    private fun File.readGroups(): List<Group> {
        return readText()
            .split("\n\n")
            .map { rawGroup ->
                val answers = rawGroup
                    .trim()
                    .split("\n")
                    .map { singlePerson ->
                        val chars = singlePerson
                            .filter { !it.isWhitespace() }
                            .toSet()
                        Answer(chars)
                    }
                Group(answers)
            }
    }

    data class Group(
        val answers: List<Answer>
    ) {

        fun countEachContains(): Int {
            val shortestAnswer = answers.minByOrNull { it.chars.size }!!
            return countContains(shortestAnswer.chars) { acc, other -> acc.intersect(other) }
        }

        fun countAnyContains(): Int {
            return countContains(setOf()) { acc, other -> acc.union(other) }
        }

        private fun countContains(initial: Set<Char>, policy: (acc: Set<Char>, other: Set<Char>) -> Set<Char>): Int {
            return answers.map { it.chars }.fold(initial, policy).size
        }
    }

    data class Answer(
        val chars: Set<Char>
    )
}

File: day6/clojure/src/day6.clj
(ns day6)

(use
 'clojure.java.io
 'clojure.set)

(defn count-occurrence [group, reductor]
  (count (reduce reductor group)))

(defn load-groups [string]
  (for [group (clojure.string/split string #"\n\n")]
    (for [answer (clojure.string/split (clojure.string/trim group) #"\n")]
      (set (clojure.string/split answer #"")))))

(defn sum-by [groups group-reductor]
  (reduce + (for [group groups] (count-occurrence group group-reductor))))

(defn -main [& args]
  (def file-contents (slurp (clojure.java.io/resource "day6.txt")))
  (def groups (load-groups file-contents))
  (println (sum-by groups clojure.set/union))
  (println (sum-by groups clojure.set/intersection)))
